Third party auth functionality for django-sudo on LMS/Studio side.

waheedahmed opened the pull request:
Added third party auth functionality on sudo pages both studio and LMS, to achieve that i created a django template tag in ```django_sudo_helpers``` and created a pipeline function to grant sudo access on successful third party re-authentication and included it into ```SOCIAL_AUTH_PIPELINE``` on startup. Also enabled third_party_auth in studio to add third party auth functionality on studio sudo page.

And the third party auth was broken in initial django-sudo PR because if third party auth user's doesn't have a password then they were not able to access instructor dashboard (LMS) and course team pages (Studio), so these changes allow user's to get sudo access on instructor dashboard (LMS) and course team page (Studio) by re-authenticating through third party auth button's.

These buttons will remain disabled if user doesn't have a third party account, and this doesn't mean that they can manually enable these buttons in browser and can use these buttons because with disabling these buttons I'm also not adding the ```pipeline_url```, so they can't use these buttons by manually enabling in browsers.

Added some unittests and bokchoy tests.

waheedahmed commented:
jenkins run python

waheedahmed commented:
@cpennington @adampalay @cahrens Please review, added 3rd party auth functionality for django-sudo on LMS side.

Sandbox: http://django-sudo.sandbox.edx.org/

@explorerleslie FYI.

cahrens commented:
I think it would make more sense for @dan-f to review this, as he reviewed the first sudo story. I will put a story in this sprint for the review.

Note that with the sudo code enabled, we were seeing that Course Team (Studio) acceptance tests were failing a lot (nearly every job). That will need to be fixed before the sudo work is merged again. See https://openedx.atlassian.net/browse/TNL-2846 and the record it is duped to.

Also, I had noticed when the sudo work was in master that I was getting server-side sudo errors when loading courses (I wasn't testing sudo at all... just saw the errors in my log). Please take a look at server-side logs to verify that there aren't errors when using courses in Studio and LMS.

waheedahmed commented:
@cahrens I recently rebased this branch with master a no test is failing now, also i can't find any error's in server-side logs too.

cahrens commented:
@waheedahmed The tests didn't fail last time before you merged the PR, but once you merged it, the Course Team acceptance tests started failing a lot. I think it is an existing sporadic issue that got much worse with the sudo logic.

waheedahmed commented:
@cahrens Then why those tests are not failing on my PR and what change is happing after merging it in master which causing failure?

cahrens commented:
I don't know... probably just timing. @jzoldak FYI that the sudo PR will be going back into master sometime in the next few weeks. @waheedahmed isn't seeing those acceptance tests failing on his branch. Do you have any recommendations, or are you OK with just waiting to see what happens once the PR is in master?

jzoldak commented:
@waheedahmed can you please temporarily decorate that entire testcase class with an instruction to run each testcase multiple times (e.g. @flaky(max_runs=5, min_passes=5), see the bottom of the page https://openedx.atlassian.net/wiki/display/TE/Flaky+Test+Process), and then trigger a build to verify that no failures happen on jenkins?

Don't merge that in to master, but it would be good to see the tests pass multiple times on jenkins.

waheedahmed commented:
jenkins run bokchoy

waheedahmed commented:
@cahrens and @jzoldak I added the ```flaky``` decorator on tests and now one test failed but after investigating it I can't find anything in django-sudo which actually affecting those tests, also in screenshot of failure ```other``` user is present on page see here [CourseTeamPageTest.test_admins_should_be_able_to_make_other_people_into_admins.png] (https://build.testeng.edx.org/job/edx-platform-test-subset/158040/artifact/test_root/log/shard_2/common.test.acceptance.tests.studio.test_studio_course_team.CourseTeamPageTest.test_admins_should_be_able_to_make_other_people_into_admins.png)
But this property [users.py#L41-L47] (https://github.com/edx/edx-platform/blob/master/common/test/acceptance/pages/studio/users.py#L41-L47) is returning an empty list of users.

And this test was already marked as flaky before last merge of django-sudo PR into master [TNL-2667] (https://openedx.atlassian.net/browse/TNL-2667), So I think django-sudo is not the reason for making it worse.

jzoldak commented:
@waheedahmed I think that wait_for_ajax_or_reload method does not do the right thing. It should be "return typeof(jQuery)!='undefined' && jQuery.active==0" rather than "return typeof(jQuery) == 'undefined' || jQuery.active == 0".

That said, I think that the wait_for_ajax_or_reload method was created to get around a bug in bok-choy's wait_for_ajax that we have subsequently fixed upstream, using the above code. The platform has already been updated to use this version (0.4.3).

Can you remove the wait_for_ajax_or_reload method and replace those calls with wait_for_ajax and see if the flakiness goes away?

waheedahmed commented:
@jzoldak sure, I'll try that.

waheedahmed commented:
jenkins run bokchoy

waheedahmed commented:
@jzoldak After using ```wait_for_ajax()``` the tests are passing with ```flaky(max_runs=10, min_passes=10)```, is it good or we need to increase the number's?

jzoldak commented:
@waheedahmed I think that's good it it's passing with 10.

waheedahmed commented:
jenkins run bokchoy

waheedahmed commented:
jenkins run python

waheedahmed commented:
@cpennington @explorerleslie @adampalay Third party auth for studio and LMS implemented.

cpennington commented on common/djangoapps/microsite_configuration/templatetags/microsite.py:
```diff
@@ -62,3 +64,38 @@ def microsite_css_overrides_file():
+
+
+@register.simple_tag(name="third_party_auth_links", takes_context=True)
+def third_party_auth_links(context):
```
Why is this part of `microsites`?

cpennington commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -647,3 +653,28 @@ def associate_by_email_if_login_api(auth_entry, backend, details, user, *args, *
+
+
+@partial.partial
+def create_sudo_session(strategy, auth_entry, *args, **kwargs):
```
I'm not wild about `third_party_auth` having to explicitly know about sudo mode. Could this method be added as part of a micro package that handles the intersection between these features, and injects the function into the `SOCIAL_AUTH_PIPELINE` on startup?

cpennington commented on lms/templates/sudo/sudo.html:
```diff
@@ -1,21 +1,24 @@
 {% extends "main_django.html" %}
-{% load i18n %}
+{% load i18n microsite %}
```
Again, this seems odd. Why do we need to pull in microsites here?

dan-f commented on cms/envs/aws.py:
```diff
@@ -357,3 +357,20 @@
+##### Third-party auth options ################################################
+if FEATURES.get('ENABLE_THIRD_PARTY_AUTH'):
+    AUTHENTICATION_BACKENDS = (
+        ENV_TOKENS.get('THIRD_PARTY_AUTH_BACKENDS', [
```
This seems like it belongs in our configuration repo, not inline in `cms/envs/aws.py`.

waheedahmed replied:
@dan-f I just copied it from ```lms/envs/aws.py``` as it is, don't know it belongs to configuration or not.

dan-f replied:
I see now. I think storing this as either configuration models or in configuration could keep it in sync with the LMS settings. The problem with copy-pasting this is that the two lists could grow out of sync. @bradenmacdonald - looks like you've worked quite a bit on third party auth - any suggestions?

bradenmacdonald replied:
Changes to this setting can result in the server loading arbitrary code, so for security it should not be configurable via configuration models. If you want to make an ansible setting that writes to ENV_TOKENS, that would make sense in terms of easing administration. Otherwise I'm not really sure how to share settings between the LMS and CMS - I think that's a well-known issue with the current codebase that has yet to be addressed, but is probably beyond the scope of this PR.

dan-f replied:
@maxrothman what do you think of having this setting synced between LMS/CMS via the configuration repo? Would it be possible to have a common setting which is written to both the cms and lms env files, and do you think it owuld make sense in this case? Pinging you since it would require coordination with devops.

dan-f commented on cms/envs/common.py:
```diff
@@ -781,6 +781,8 @@
     # Allows sudo-mode
     'sudo',
+
+    'microsite_configuration'
```
Why does this have to be an installed app?

dan-f commented:
@waheedahmed are you planning to add unit and acceptance tests?

waheedahmed commented:
jenkins run_a11y

waheedahmed commented:
jenkins run a11y

explorerleslie commented:
Waheed, is the latest code on the sandbox? django-sudo.sandbox. I'd like to
have another look.

On Tue, Aug 25, 2015 at 7:42 AM, Waheed Ahmed <notifications@github.com>
wrote:

> jenkins run a11y
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/9180#issuecomment-134561359>.
>

waheedahmed commented:
@explorerleslie Not yet let me update.

waheedahmed commented:
@explorerleslie Sandbox is updated with latest code.

waheedahmed commented:
jenkins run all

explorerleslie commented:
I'm getting ERR_NAME_NOT_RESOLVED when trying to hit your sandbox, @waheedahmed

waheedahmed commented:
@explorerleslie Looks like old sandbox was destroyed, lemme create a new one. I'll let you know when its done.

waheedahmed commented:
@explorerleslie Sandbox is up now with Google and Facebook authentication enabled.

waheedahmed commented:
jenkins run python

waheedahmed commented:
jenkins run bokchoy

waheedahmed commented:
jenkins run python

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_startup.py:
```diff
@@ -0,0 +1,30 @@
+    Test django-sudo startup
+    """
+
+    def setUp(self):
```
Please remove this empty `setUp`.

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_startup.py:
```diff
@@ -0,0 +1,30 @@
+    def setUp(self):
+        super(DjangoSudoStartupTestCase, self).setUp()
+
+    @patch.dict("django.conf.settings.FEATURES", {"ENABLE_THIRD_PARTY_AUTH": False})
```
Try using from `django.test.utils.override_settings` for this type of thing.

adampalay replied:
@dan-f , I think that's normally the right thing to do, but for `FEATURES`, we don't want to override the whole dictionary, we just want to override the one key-value pair in `FEATURES`.

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_startup.py:
```diff
@@ -0,0 +1,30 @@
+            self.assertFalse(mock_third_party_for_sudo.called)
+
+    @patch.dict("django.conf.settings.FEATURES", {"ENABLE_THIRD_PARTY_AUTH": True})
+    def test_run_without_theme(self):
```
Could you squash these two tests into a `ddt` test case?

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_third_part_auth_links.py:
```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
```
Typo in this file name: `part` should be `party`.

dan-f commented:
@waheedahmed could you please add a cover letter or description to the initial comment on this PR summarizing the changes you've made and how they address the problem of 3rd party auth being broken in the initial django_sudo PR?

waheedahmed commented:
@dan-f @cpennington PR is ready for review with third party auth feature enabled, also added some unittests and bokchoy tests.

dan-f commented:
@waheedahmed thank you for the description!

dan-f commented on cms/envs/bok_choy.env.json:
```diff
@@ -103,5 +104,12 @@
     "TIME_ZONE": "America/New_York",
-    "WIKI_ENABLED": true
+    "WIKI_ENABLED": true,
+    "THIRD_PARTY_AUTH_BACKENDS": [
```
Are any of the backends in this list actually used in bok_choy tests other than the dummy backend?

waheedahmed replied:
@dan-f Nope, I'll remove other backends.

dan-f commented on cms/envs/devstack.py:
```diff
@@ -118,3 +118,8 @@ def should_show_debug_toolbar(_):
+########################## Third Party Auth #######################
+
+if FEATURES.get('ENABLE_THIRD_PARTY_AUTH') and 'third_party_auth.dummy.DummyBackend' not in AUTHENTICATION_BACKENDS:
+    AUTHENTICATION_BACKENDS = ['third_party_auth.dummy.DummyBackend'] + list(AUTHENTICATION_BACKENDS)
```
I really don't like copying and pasting settings across lms/cms. I understand that we currently have no shared common settings file in which to place this stuff, so we may have to live with this if configuration models don't work out.

marcotuts commented:
@frrrances  - created a UX ticket, added to next sprint and tentatively assigned to you. (https://openedx.atlassian.net/browse/UX-2763)

dan-f commented on common/djangoapps/django_sudo_helpers/pipeline.py:
```diff
@@ -0,0 +1,38 @@
+        match = COURSE_REGEX.match(redirect_to)
+        if not match:
+            # COURSE_REGEX will not work for studio
+            course_regex = re.compile(r'^.*?/course_team/{}'.format(settings.COURSE_ID_PATTERN))
```
Why not just match against `settings.COURSE_ID_PATTERN`? Why is it necessary to include `/course_team/` in the regex?

waheedahmed replied:
@dan-f I don't think it will work because we are matching ```course_id``` from url, consider this if url is like this ```/course_team/edx/demo/2015_Q2``` and we match against ```settings.COURSE_ID_PATTERN``` then it will return this ```course_team/edx/demo``` instead of ```edx/demo/2015_Q2``` thats why I'm including ```/course_team/``` in regex like ```/courses/``` in ```COURSE_REGEX```.

dan-f replied:
I see. If we want to be certain that we're matching against valid course_ids, we could match the `COURSE_REGEX`/`COURSE_ID_PATTERN` against the URL, and then try to load each match using `CourseKey.from_string()`. Whichever one is valid as a course key would be the region. Then we don't have to special-case certain URLs, which will be difficult to maintain.

The performance cost of that approach might not be great, but since this method is unlikely to be called often, I don't think we'd have to worry about it... @cpennington does this concern you?

adampalay replied:
@dan-f that sounds really dangerous. This may not be the most elegant solution, but it's way less risky

dan-f replied:
@adampalay are you referring to performance risk or some other kind of risk?

adampalay replied:
like if you had a course whose org was "course_team"

dan-f commented on cms/templates/sudo/third_party_auth_links.html:
```diff
@@ -0,0 +1,17 @@
+<div class="form-actions form-third-party-auth">
+
+{% for enabled in providers %}
+    <button type="submit" class="button button-primary button-{{enabled.provider_id}} login-{{enabled.provider_id}}" {% if enabled.has_account %} onclick="thirdPartySignin(event, '{{pipeline_url|get_pipeline_url:enabled.provider_id}}');" {% else %} disabled="disabled" {% endif %} ><span class="icon fa {{enabled.icon_class}}"></span>{% trans 'Sign in with' %} {% trans enabled.name %}</button>
```
Concatenating translated strings like this doesn't provide sufficient context to our translators. Please check out our i18n guide on how to design translated strings with proper context: https://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#don-t-append-strings-interpolate-values

dan-f commented on lms/templates/sudo/third_party_auth_links.html:
```diff
@@ -0,0 +1,17 @@
+{% load i18n third_party_auth_links %}
```
Can this template and the other one be shared between Studio and LMS by keeping it in the `django_sudo_helpers` app?

dan-f commented on lms/templates/sudo/sudo.html:
```diff
@@ -1,21 +1,24 @@
+    {% third_party_auth_links %}
+
 </section>
 {% endblock %}
```
Could this content (duplicated in `cms/templates/sudo/sudo.html`) also be shared in the `sudo_helpers` app and included with template tags?

waheedahmed replied:
@dan-f We can't move this to ```dajngo_sudo_helpers``` because this is not used in ```dajngo_sudo_helpers```, its being loaded in third party module https://github.com/mattrobenolt/django-sudo. We need to place it in root of template dir like this ```templates/sudo/sudo.html```.

waheedahmed replied:
@dan-f sorry my bad, I'll do that.

waheedahmed replied:
@dan-f We can only move ```third_party_auth_links.html``` into ```django_sudo_helpers``` because the content of both ```sudo.html```'s is different from each other except ```{% load i18n third_party_auth_links %}``` and ```{% third_party_auth_links %}```.

dan-f replied:
Got it. Thanks.

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_startup.py:
```diff
@@ -0,0 +1,29 @@
+        True,
+        False,
+    )
+    def test_run_third_party_auth(self, is_enable):
```
Nit: typo - `is_enable` should be `is_enabled`

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_third_party_auth_links.py:
```diff
@@ -0,0 +1,58 @@
+
+
+@unittest.skipUnless(testutil.AUTH_FEATURE_ENABLED, 'third_party_auth not enabled')
+class ThirdPartyAuthLinkTests(testutil.TestCase, TestCase):
```
No need to inherit from `django.test.TestCase`, as `testutil.TestCase` already does this.

dan-f commented on common/djangoapps/django_sudo_helpers/tests/test_third_party_auth_links.py:
```diff
@@ -0,0 +1,58 @@
+
+        # Assert that button classes are present in response for enabled providers.
+        self.assertIn('button-oa2-google-oauth2 login-oa2-google-oauth2', response)
+        self.assertIn('button-oa2-linkedin-oauth2 login-oa2-linkedin-oauth2', response)
```
Since we're testing that the proper links show up, I think it would be better to assert that the expected URLs show up rather than their associated button classes.

dan-f commented on common/test/acceptance/pages/studio/users.py:
```diff
@@ -8,19 +8,6 @@
 from . import BASE_URL
 
 
-def wait_for_ajax_or_reload(browser):
```
Why did you delete this method and replace all of its invocations with `wait_for_ajax`?

waheedahmed replied:
@dan-f This method was making a test flaky, I removed it after discussing with @jzoldak, see our above conversation.

dan-f replied:
Ah I missed that.

dan-f commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -444,3 +447,50 @@ def test_pending_tasks_section_is_visible(self):
+        sudo_password_page = SudoPage(self.browser, self.instructor_dashboard_page)
+        sudo_password_page.visit()
+        sudo_password_page.click_third_party_dummy_provider_button()
+        self.instructor_dashboard_page.visit()
```
Correct me if I'm wrong, but doesn't this line guarantee that this test will never fail? I expected that after clicking on the dummy provider button, we should assert that we are now on the instructor dashboard page. By manually calling `visit()` we never check if the redirect worked.

waheedahmed replied:
@dan-f Actually don't need to manually call visit on page because when we are clicking on dummy provider button it's actually already waiting for that page. So we don't need to assert and visit both.

dan-f commented on common/test/acceptance/tests/studio/test_studio_course_team.py:
```diff
@@ -352,3 +356,46 @@ def test_admins_should_be_able_to_give_course_ownership_to_someone_else(self):
+        sudo_password_page.visit()
+        sudo_password_page.click_third_party_dummy_provider_button()
+
+    def _link_dummy_account(self):
```
Please put this method in a common place - we should avoid copying/pasting whenever possible!

dan-f commented on common/test/acceptance/tests/studio/test_studio_course_team.py:
```diff
@@ -352,3 +356,46 @@ def test_admins_should_be_able_to_give_course_ownership_to_someone_else(self):
+        # Visit sudo page and click on dummy auth button to get sudo access.
+        sudo_password_page = SudoPage(self.browser, self.page)
+        sudo_password_page.visit()
+        sudo_password_page.click_third_party_dummy_provider_button()
```
Same note here about asserting that we're actually on the page.

dan-f replied:
I now see that the assertion is hidden in the `wait_for_page()` call in `click_third_party_dummy_provider_button()`. I think that assertions belongs here rather than buried in a PageObject method.

dan-f commented:
@waheedahmed I finished my first pass of review. Please let me know when you've responded to my comments so that I can re-review. Thanks!

dan-f commented:
One more thing - I imagine this will need to be vetted for accessibility. Please make sure that you've followed our [accessibility guidelines](https://edx.readthedocs.org/projects/edx-developer-guide/en/latest/accessibility.html) and then see if @cptvitamin needs to review for accessibility.

bradenmacdonald commented on common/djangoapps/django_sudo_helpers/settings.py:
```diff
@@ -0,0 +1,8 @@
+
+def apply_settings(django_settings):
+    """Set provider-independent settings."""
+    django_settings.SOCIAL_AUTH_PIPELINE += ('django_sudo_helpers.pipeline.create_sudo_session',)
```
The exact ordered list of what steps are in the pipeline is quite important to security of third_party_auth, and I don't really like having the pipeline setting being sometimes modified by code elsewhere in the codebase. I'd prefer that you put `if using_django_sudo: django_settings.SOCIAL_AUTH_PIPELINE += ('django_sudo_helpers.pipeline.create_sudo_session',)` right into `third_party_auth/settings.py` where the rest of the pipeline steps are all configured, or at least add a comment there explaining that sometimes an extra pipeline step will be added.

waheedahmed replied:
@bradenmacdonald First i did this in ```third_party_auth/settings.py``` but @cpennington asked me if we can move it into ```django_sudo_helpers``` and inject this into ```SOCIAL_AUTH_PIPELINE``` on startup.

I'll add a comment explaining this.

bradenmacdonald replied:
haha, sorry for the conflicting feedback then. A comment in the other file works for me!

waheedahmed commented on common/djangoapps/third_party_auth/settings.py:
```diff
@@ -43,6 +43,7 @@ def apply_settings(django_settings):
 
     # Inject our customized auth pipeline. All auth backends must work with
     # this pipeline.
+    # Sometimes an extra pipeline step will be added on startup for django-sudo, if enabled.
```
@bradenmacdonald FYI.

bradenmacdonald replied:
Thanks!

waheedahmed commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -150,6 +151,7 @@ def is_api(auth_entry):
 
     AUTH_ENTRY_LOGIN_API,
     AUTH_ENTRY_REGISTER_API,
+    AUTH_ENTRY_SUDO
```
@bradenmacdonald Can we remove frozenset and use only set to do the same thing like pipeline?

dan-f replied:
Whichever way you choose, please be consistent between this and the pipeline settings

bradenmacdonald replied:
@waheedahmed Yes you can, though I'm not a big fan of modifying an enum definition at runtime. Personally I'd much rather see you define the string constant in this file as is done for all the other constants in this enum. `AUTH_ENTRY_SUDO` is just a constant and its actual value doesn't matter, so there's no need to import the django_sudo_helpers app in order to define the constant, and there's no harm in having it always defined. (The pipeline case we had discussed is somewhat different because there you are injecting a code dependency.)

If you do take the runtime modification approach, please add a comment explaining it.

waheedahmed replied:
Thanks @bradenmacdonald!

waheedahmed commented:
@dan-f, @bradenmacdonald Comments addressed.

waheedahmed commented:
jenkins run lettuce

explorerleslie commented:
just filled @dan-f in on the history here of why we need third party auth in Studio for this case, so that comment is resolved.

explorerleslie commented:
:+1: I would have @carsongee also take a look to make sure this fulfills the MIT third-party auth needs.

waheedahmed commented:
@cptvitamin Accessibility review please.

dan-f commented on common/djangoapps/django_sudo_helpers/templates/sudo/third_party_auth_links.html:
```diff
@@ -0,0 +1,17 @@
+    }
+</script>
+<span class="deco-divider" style="background: transparent;">
+    <span class="copy">{% trans "or" %}</span>
```
This `trans` call does not give sufficient context to our human translators - please see https://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html

dan-f commented on common/djangoapps/django_sudo_helpers/templates/sudo/third_party_auth_links.html:
```diff
@@ -0,0 +1,17 @@
+<div class="form-actions form-third-party-auth">
+
+{% for enabled in providers %}
+    <button type="submit" class="button button-primary button-{{enabled.provider_id}} login-{{enabled.provider_id}}" {% if enabled.has_account %} onclick="thirdPartySignin(event, '{{pipeline_url|get_pipeline_url:enabled.provider_id}}');" {% else %} disabled="disabled" style="color: #ffffff;" {% endif %} ><span class="icon fa {{enabled.icon_class}}"></span>{% trans 'Sign in with' %} {% trans enabl ...
```
The `trans` calls on this line will also cause problems for our translators.

1) They will receive the message "Sign in with" which doesn't give them any context. Sign in with what? I suggest using a format string so that they can see something like "Sign in with {provider_name}". See [this part](https://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#don-t-append-strings-interpolate-values) of the i18n guide.

2) The `trans` call around the `enabled.name` value will not work unless `enabled.name` has been passed into a  `uggettext_noop` call earlier.

cptvitamin replied:
please hide the icon from visual screen readers with aria-hidden=true.

dan-f commented:
@waheedahmed thanks for addressing my feedback. Once you work out the course_id matching and address the translation comments I made, this gets a :+1: from me :)

I've heard from UX that they plan to review this in their next sprint.

cptvitamin commented on lms/templates/sudo/sudo.html:
```diff
@@ -1,21 +1,24 @@
-        <div style="margin: 0px auto; width: 218px;">
+        <div style="margin: 0px auto; width: 320px;">
             <form class="sudo-form" method="post">{% csrf_token %}
             {{ form.as_p }}
```
where is this form?

waheedahmed replied:
@cptvitamin This form is passed by a third party module to this template, https://github.com/edx/django-sudo/blob/master/sudo/forms.py#L13-L28

frrrances replied:
Can we add a div around this with a class of "field" to use default input and label styles in _account.scss ?

cptvitamin commented on cms/templates/sudo/sudo.html:
```diff
@@ -14,9 +14,11 @@
         <form class="sudo-form" method="post" action="">{% csrf_token %}
             {{ form.as_p }}
```
where is this form?

waheedahmed replied:
@cptvitamin Same here. https://github.com/edx/django-sudo/blob/master/sudo/forms.py#L13-L28

carsongee commented:
I'm afraid I'm unlikely to find the time to look at this.  @bdero or @pdpinch, you will want to make sure this still works for both our authentication scenarios.  I don't see anything about CAS or SSL authentication in the diff, so think it may not work at first glance, but I'm not sure.

waheedahmed commented:
jenkins run bokchoy

adampalay commented:
@explorerleslie and @carsongee , I don't think this will fulfill MIT's needs ( @waheedahmed , is that right?), which is why we're feature-flagging it.

explorerleslie commented:
@adampalay oh, right. carry on.

waheedahmed commented:
@adampalay right, here is its PR:  https://github.com/edx/edx-platform/pull/9552

@carsongee do you guys have different login pages for CAS and SSL authentication?

waheedahmed commented:
@dan-f Are we good now?

dan-f commented:
:+1:

cpennington commented:
:+1:

waheedahmed commented:
@marcotuts, @frrrances can you please give an expected date for UX review?

frrrances commented on lms/templates/sudo/sudo.html:
```diff
@@ -1,21 +1,25 @@
+        <div style="margin: 0px auto; width: 320px;">
             <form class="sudo-form" method="post">{% csrf_token %}
             {{ form.as_p }}
             <p>
```
Please add "field" class to this p tag to use default input and label styles in _account.scss

frrrances commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -755,3 +755,30 @@
     background: tint($red,95%);
   }
 }
+
```
Remove these added styles and use the existing ones above by adding classes.

frrrances commented on cms/templates/sudo/sudo.html:
```diff
@@ -1,7 +1,8 @@
+    {% load i18n third_party_auth_links %}
     {% load compressed %}
     {% compressed_css 'style-main' %}
+    {% compressed_css 'style-vendor' %}
```
style-vendor always needs to come before style-main. please move this above style-main.

frrrances commented on cms/templates/sudo/sudo.html:
```diff
@@ -1,7 +1,8 @@
 {% block body %}
```
This template is rendering as a partial in my browser... it's missing the core html structure of html, head, body tags. Also visually it is in nowhere land. If you are going to do this in a modal, can you have it render over a blank Studio page? Or can you do the same as you did in the LMS and render the header footer?

waheedahmed replied:
@frrrances Studio is using mako templates but sudo template is rendered in django templates and we don't have a main layout in django templates for studio but we have it for LMS, that's why I added this modal after discussing it with @explorerleslie and @cpennington.

We can't add studio header/footer.

cpennington replied:
Well, we could. We'd just need to make a copy of the studio base template and translate it to django, like we did with the LMS template.

waheedahmed replied:
@cpennington I think we discussed it before and it wasn't possible that's why we decided not to do that.

cpennington replied:
Honestly, I don't recall at this point whether we discussed that. Do you remember why that wasn't going to be possible?

waheedahmed replied:
@cpennington I can't recall it either but I asked you question about it on my original django-sudo PR here https://github.com/edx/edx-platform/pull/7404#issuecomment-95612166 and for some reason I never converted ```base.html``` into django template.

I don't remember exactly why we never converted it.

cpennington replied:
Hrm... My response to that comment makes it seem like I was in favor of a django version as well...

frrrances commented on common/djangoapps/django_sudo_helpers/templates/sudo/third_party_auth_links.html:
```diff
@@ -0,0 +1,23 @@
+<div class="form-actions form-third-party-auth">
+
+{% for enabled in providers %}
+    <button type="submit" class="button button-primary button-{{enabled.provider_id}} login-{{enabled.provider_id}}" {% if enabled.has_account %} onclick="thirdPartySignin(event, '{{pipeline_url|get_pipeline_url:enabled.provider_id}}');" {% else %} disabled="disabled" style="color: #ffffff;" {% endif %} ><span aria-hidden="true" class="icon fa {{enabled.icon_class}}"></span>
```
Remove inline styles.

frrrances replied:
It would be nice to remove the inline js as well.

frrrances commented on cms/static/sass/elements/_modal.scss:
```diff
@@ -181,7 +181,91 @@
         text-shadow: 0px 1px 0px #F9F8F8;
         background-clip: padding-box;
         font-size: 0.8125em;
+        width: 100%;
```
Remove this new code from the modals partial when you move the form into the page frame.

frrrances commented on cms/static/sass/elements/_typography.scss:
```diff
@@ -23,6 +23,10 @@
  font-weight: 200;
 }
 
+%t-semibold {
```
Use existing values above.

frrrances commented:
@waheedahmed Sorry for the delay in reviewing. This was a lot of code and I needed to get some context about what this feature was. It seems like UX wasn't tagged on the original add sudo PR yet, so I have some general concerns which I will add here. I have the following concerns that need to be addressed before this can merge: 

- [ ] There is a lot of duplicated code, especially sass, in here that could easily use existing styles. I have noted most of the places that you can add classes to your HTML and have the styles for free. (You can look at the HTML on the LMS and Studio login pages to see how the DOM and classes are being structured.)

- [ ] Please use existing color, type, layout variables and mixins, and do not create styles inline! If you are unsure how to write the sass with what exists, please put it in the _developer.scss file so we can make sure it is theme-able and not creating excess css. If you are comfortable with sass (using variables, mixins, etc) and need to write custom sass for the sudo views, please add a new _sudo.scss partial and add it in the appropriate place in the import.

- [ ] For LMS - the form is okay, but it would be better to have the header use the course context rather than the generic header. Also, if the social buttons are disabled, it would be better to not show them.

- [ ] For Studio - the form should not be in a modal and should not use LMS modal styles. Please move the form into the Studio body (like the LMS form) and make it look more like the Studio login page (see example below).

![studio-sudo-access](https://cloud.githubusercontent.com/assets/4327102/10145814/6649bd5c-65f3-11e5-9aa6-5b668b0b82b2.png)

Also, here are the various states for the social buttons: 

Default: 
![default-social](https://cloud.githubusercontent.com/assets/4327102/10145776/2e387b56-65f3-11e5-8114-8fa5466d5fba.png)

Hover:
![hover-social](https://cloud.githubusercontent.com/assets/4327102/10145777/2e3ba452-65f3-11e5-91fc-5dd8cb6c87c4.png)

Disabled:
![disabled-social](https://cloud.githubusercontent.com/assets/4327102/10145778/2e3b97c8-65f3-11e5-9b1d-33cc3fc4feec.png)

Asking for a user's password at an unusual place in their workflow is a sensitive moment and we want the experience to be as much like the surrounding app as possible to make the user feel comfortable continuing. Let me know if you have questions and when this PR is ready for another review.

waheedahmed commented:
@frrrances Comments addressed please take another look.

If we want add course context header for LMS/Studio then we need to convert some templates into django from mako.

waheedahmed commented:
@frrrances Also removed modal on Studio side.

<img width="1272" alt="screen shot 2015-10-01 at 7 26 35 pm" src="https://cloud.githubusercontent.com/assets/2851134/10224812/3924ee54-6879-11e5-84c5-521e567af3ef.png">

waheedahmed commented:
jenkins run bokchoy

waheedahmed commented:
@frrrances sandbox updated with latest code.

frrrances commented:
@waheedahmed Thank you for the updates. I didn't follow the conversation with @cpennington about django/mako templates - is it a big deal to convert the templates? If not, are there concerns or issues to doing that (duplicate code, performance issues, etc)? It would be much better user experience to have the background stay in context so I would like to understand the challenges for doing that before I give final thumbs. I will review the rest of the changes today.

waheedahmed commented:
@frrrances, @cpennington I successfully converted the studio ```base.html``` into django template, also added headers with course context on both sides LMS/CMS.

Sandbox is updated with latest code.

<img width="1278" alt="screen shot 2015-10-05 at 8 14 53 pm" src="https://cloud.githubusercontent.com/assets/2851134/10286412/008bf480-6ba7-11e5-86f9-734e6c02d2ad.png">

frrrances commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -491,7 +491,7 @@
     display: block;
     margin: ($baseline*1.5) 0;
-    border-top: ($baseline/5) solid $m-gray-l4;
+    border-top: ($baseline/5) solid white;
```
Use $white variable.

cpennington commented on cms/templates/base_django.html:
```diff
@@ -0,0 +1,77 @@
+        {% if user.is_authenticated %}
+            {% include "widgets/sock.html" with online_help_token=None %}
+        {% endif %}
+        {% include "widgets/footer.html" %}
```
Are these django or mako widget templates?

waheedahmed replied:
@cpennington these are mako templates.

cpennington replied:
Ok, good.

frrrances commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -500,7 +500,7 @@
       left: 43%;
       padding: ($baseline/4) ($baseline*1.5);
-      background: white;
+      background: #f2f2f2;
```
Use $body-bg variable

frrrances commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -755,3 +755,13 @@
   }
 }
+
+#sudo {
```
Don't style IDs - I think you can switch this to a class selector (.sudo) and have it work.

frrrances commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -755,3 +755,13 @@
+
+#sudo {
+  margin: 0 auto;
+  width: 480px;
```
Switch to "max-width"

cpennington commented:
I'm :+1: on this.

frrrances commented on cms/templates/sudo/sudo.html:
```diff
@@ -1,42 +1,35 @@
+{% block content %}
+
+<div class="wrapper-content wrapper">
+    <section aria-hidden="true" class="content view-sudo">
```
Move the "view-sudo" class to the bodyclass (line 9)

frrrances commented on cms/templates/sudo/sudo.html:
```diff
@@ -1,42 +1,35 @@
+{% compressed_css 'style-main' %}
+
+{% block title %}{% trans "Confirm Your Password" %}{% endblock %}
+{%block bodyclass %} is-signedin course view-outline {% endblock %}
```
Remove "view-outline" class and replace with "view-sudo"

frrrances commented on cms/static/sass/views/_account.scss:
```diff
@@ -243,3 +244,118 @@
+      width: flex-grid(8, 8);
+      margin-bottom: $baseline;
+
+      button[type="submit"] {
```
use a class selector here - "action-primary"

frrrances commented on common/djangoapps/django_sudo_helpers/templates/sudo/third_party_auth_links.html:
```diff
@@ -0,0 +1,25 @@
+      window.location.href = url;
+    }
+</script>
+<span class="deco-divider">
```
Move this decorator inside the if enabled block (line 16) so it doesn't show if providers are not enabled.

frrrances commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -755,3 +755,13 @@
+
+  h2 {
+    text-align: center;
+  }
```
Add rule after h2 to style error message: 
```
.errorlist {
    border-left: 3px solid $error-color;
    padding: 10px 15px;
    background-color: $gray-l4;
    list-style-type: none;
}
```

frrrances commented on cms/static/sass/views/_account.scss:
```diff
@@ -243,3 +244,118 @@
+    h1 {
+      text-align: center;
+    }
+
```
Add rule to add styling for error messages: 
```
.errorlist {
    border-left: 3px solid $error-color;
    padding: 10px 15px;
    background-color: $gray-l6;
    list-style-type: none;
}
```

frrrances commented:
@waheedahmed Thanks for making the changes - this is looking much better. I had a few small code comments inline, and just two small general changes (which I also marked in the code): 

1. Add some error styles to emphasize the error when the password is wrong so the user notices it (I added the style rule inline)
2. Hide the " -- or -- " if 3rd party isn't enabled. 

The sandbox doesn't seem to have 3rd party enabled. Is there a way I can double check those buttons? Once I can take a look at those, this looks good to go.

waheedahmed commented:
@frrrances Addressed your comments and sandbox is updated with the latest code.

Third party auth is enabled on sandbox but you can't login through third party on Studio side but you can see third party auth buttons on LMS login page and you can login through them. But on Studio side you need to link an account from LMS side and set edX password for it and then login with username and password on studio side then you can see third party auth buttons on Studio sudo page.

frrrances commented:
@waheedahmed Code and UX look good. Thanks! :+1:

waheedahmed commented:
jenkins run bokchoy

waheedahmed commented:
@frrrances Thanks!
